applications/conversations/controllers/api/ConversationsApiController.php

```diff
@@ -0,0 +1,449 @@
+        UserModel $userModel
+    ) {
+        $this->caseScheme = new CapitalCaseScheme();
+        $this->config = $config;
+        $this->conversationModel = $conversationModel;
+        $this->userModel = $userModel;
+    }
+
+    /**
+     * Check that the user has moderation rights over conversations.
+     *
+     * @throw Exception
+     */
+    private function checkModerationPermission() {
+        if (!$this->config->get('ConversationMessages.Moderation.Allow', false)) {
+            throw permissionException();
+        }
+        $this->permission('ConversationMessages.Moderation.Manage');
+    }
+
+    /**
+     * Get a conversation by its numeric ID.
+     *
+     * @param int $id The conversation ID.
+     * @param int|null $viewingUserID The user viewing the conversation. Should only be set if user is part of the conversation.
+     * @throws NotFoundException if the conversation could not be found.
+     * @return array
+     */
+    private function conversationByID($id, $viewingUserID = null) {
+        if ($viewingUserID) {
+            $options = ['viewingUserID' => $viewingUserID];
+        } else {
+            $options = [];
+        }
+
+        $conversation = $this->conversationModel->getID($id, DATASET_TYPE_ARRAY, $options);
+        if (!$conversation) {
+            throw new NotFoundException('Conversation');
+        }
+
+        return $conversation;
+    }
+
+//
+//    Uncomment and test once ConversationModel::delete() is properly implemented.
+//    See https://github.com/vanilla/vanilla/issues/5897 for details.
+//
+//    /**
+//     * Delete a conversation.
+//     *
+//     * @param int $id The ID of the conversation.
+//     * @≠=throws NotFoundException if the conversation could not be found.
+//     * @throws MethodNotAllowedException if Conversations.Moderation.Allow !== true.
+//     */
+//    public function delete($id) {
+//        if (!$this->config->get('Conversations.Moderation.Allow', false)) {
+//            throw new MethodNotAllowedException();
+//        }
+//
+//        $this->permission('Conversations.Moderation.Manage');
+//
+//        $this->schema(['id:i' => 'The conversation ID'])->setDescription('Delete a conversation.');
+//        $this->schema([], 'out');
+//
+//        $this->conversationByID($id);
+//        $this->conversationModel->deleteID($id);
+//
+//        return new Data([], 204);
+//    }
+
+
+    /**
+     * Leave a conversation.
+     *
+     * @param int $id The ID of the conversation.
+     * @throws NotFoundException if the conversation could not be found.
+     * @return Data
+     */
+    public function delete_leave($id) {
+        $this->permission('Garden.SignIn.Allow');
+
+        $this->idParamSchema()->setDescription('Leave a conversation.');
+        $this->schema([], 'out');
+
+        $this->conversationByID($id);
+
+        $this->conversationModel->clear($id, $this->getSession()->UserID);
+
+        return new Data([], 204);
```

Comment:
I would go for new `Data(null, 204);` then. It's nice when you know that returning nothing means it will do a 204 for post but this is something that is hard to figure out.
Explicit return helps to know that a delete should return a 204 on success.

Reply:
We should either always return Data objects or never return them.